codes/csharp/chapter_hashing/hash_map_chaining.cs

```diff
@@ -0,0 +1,146 @@
+﻿/**
+* File: hash_map_chaining.cs
+* Created Time: 2023-06-26
+* Author: hpstory (hpstory1024@163.com)
+*/
+
+namespace hello_algo.chapter_hashing;
+
+/* 链式地址哈希表 */
+class HashMapChaining {
+    int size; // 键值对数量
+    int capacity; // 哈希表容量
+    double loadThres; // 触发扩容的负载因子阈值
+    int extendRatio; // 扩容倍数
+    List<LinkedList<Pair>> buckets; // 桶数组
+
+    /* 构造方法 */
+    public HashMapChaining() {
+        size = 0;
+        capacity = 4;
+        loadThres = 2 / 3.0;
+        extendRatio = 2;
+        buckets = new List<LinkedList<Pair>>(capacity);
+        for (int i = 0; i < capacity; i++) {
+            buckets.Add(new LinkedList<Pair>());
+        }
+    }
+
+    /* 哈希函数 */
+    private int hashFunc(int key) {
+        return key % capacity;
+    }
+
+    /* 负载因子 */
+    private double loadFactor() {
+        return (double)size / capacity;
+    }
+
+    /* 查询操作 */
+    public string get(int key) {
+        int index = hashFunc(key);
+        // 遍历桶，若找到 key 则返回对应 val
+        foreach (Pair pair in buckets[index]) {
+            if (pair.key == key) {
+                return pair.val;
+            }
+        }
+        // 若未找到 key 则返回 null
+        return null;
+    }
+
+    /* 添加操作 */
+    public void put(int key, string val) {
+        // 当负载因子超过阈值时，执行扩容
+        if (loadFactor() > loadThres) {
+            extend();
+        }
+        int index = hashFunc(key);
+        // 遍历桶，若遇到指定 key ，则更新对应 val 并返回
+        foreach (Pair pair in buckets[index]) {
+            if (pair.key == key) {
+                pair.val = val;
+                return;
+            }
+        }
+        // 若无该 key ，则将键值对添加至尾部
+        buckets[index].AddLast(new Pair(key, val));
+        size++;
+    }
+
+    /* 删除操作 */
+    public void remove(int key) {
+        int index = hashFunc(key);
+        // 遍历桶，从中删除键值对
+        LinkedListNode<Pair> head = buckets[index].First;
+        while (head != null) {
```

Comment:
I use linked lists to implement this part. This is some different from the implementation of other languages.

Reply:
Add I didn't add Pair class to hash_map_chaining and hash_map_open_addressing, because in C# cannot allowed classes with the same name under a single namespace. I'm not sure if the docs needs to be updated. If you have any suggestions, please leave a message and I will update PR.

Reply:
> I use linked lists to implement this part. This is some different from the implementation of other languages.

Better to use dynamic array for the consistency.

Reply:
> And I didn't add Pair class to hash_map_chaining and hash_map_open_addressing, because in C# cannot allowed classes with the same name under a single namespace. I'm not sure if the docs needs to be updated. If you have any suggestions, please leave a message and I will update PR.

Thanks for pointing out the problem. I was thinking make the `Pair` class as a public module. No need more action and I will handle it.

Reply:
> > I use linked lists to implement this part. This is some different from the implementation of other languages.
> 
> Better to use dynamic array for the consistency.

fixed